packages/type-utils/src/TypeOrValueSpecifier.ts

```diff
@@ -129,14 +137,26 @@ function specifierNameMatches(type: ts.Type, name: string | string[]): boolean {
 
 function typeDeclaredInFile(
   relativePath: string | undefined,
+  excludePaths: string | string[] | undefined,
   declarationFiles: ts.SourceFile[],
   program: ts.Program,
 ): boolean {
   if (relativePath === undefined) {
     const cwd = program.getCurrentDirectory().toLowerCase();
-    return declarationFiles.some(declaration =>
-      declaration.fileName.toLowerCase().startsWith(cwd),
-    );
+    const excludePathsArray =
+      excludePaths === undefined
+        ? ['node_modules']
+        : Array.isArray(excludePaths)
+        ? excludePaths
+        : [excludePaths];
+
+    return declarationFiles.some(declaration => {
+      const fileName = declaration.fileName.toLowerCase();
+      return (
+        fileName.startsWith(cwd) &&
+        excludePathsArray.every(p => !fileName.startsWith(path.join(cwd, p)))
+      );
+    });
   }
   const absolutePath = path
     .join(program.getCurrentDirectory(), relativePath)
```

Comment:
Just to point this out - you are **not** disallowing files in the `excludesPath` if they are explicitly reference in `path`. I actually think they should be disallowed as well - that would make the specifier more consistent.

See https://github.com/typescript-eslint/typescript-eslint/issues/6819#issuecomment-1500269177

Reply:
Looking closer I see that the `FileSpecifier` type disallows adding both `path` and `excludePaths` at the same time - so this wouldn't be such an issue - except for the situation where you leave `excludePaths` empty (so it defaults to `node_modules`) and add `path: "node_modules/..."` - then I think that would still actually match, right?

Reply:
`excludedPaths` is only used when `path` is `undefined` so that should be an issue.